File: tokenizers/bpe.py
"""
Heavily inspired from https://huggingface.co/learn/llm-course/en/chapter6/5
"""

from collections import Counter
from time import perf_counter
from dataclasses import dataclass
from typing import Iterator

import structlog
from bidict import bidict
from tqdm import tqdm

from tokenizers import AbstractTokenizer, TokenID, Token

log = structlog.get_logger()

Word = str

Splits = dict[Word, list[str]]
Pair = tuple[str, str]
Merges = dict[Pair, str]

APPROXIMATE_VOCAB_SIZE = False


def compute_pair_freqs(splits: Splits, freqs: Counter) -> Counter:
    pair_freqs = Counter()
    for word, freq in freqs.items():
        split = splits[word]
        if len(split) == 1:
            continue
        for i in range(len(split) - 1):
            pair = (split[i], split[i + 1])
            pair_freqs[pair] += freq
    return pair_freqs


def get_words(line: str) -> list[str]:
    line = line.strip()
    line = "".join(e for e in line.lower() if e.isalnum() or e.isspace())
    return [w for w in line.split(" ") if len(w) > 0]


@dataclass
class Tokenizer(AbstractTokenizer):
    merges: Merges
    tokens: bidict[int, str]

    def __str__(self):
        return f"Tokenizer with {len(self.tokens)} tokens and {len(self.merges)} merge rules"

    def tokenize(self, line: str) -> list[TokenID]:
        words = get_words(line)
        if len(words) == 0:
            return []
        splits: list[list[str]] = list(map(list, words))
        # Apply the merge rules in order
        for pair, merge in self.merges.items():
            for split in splits:
                i = 0
                while i < len(split) - 1:
                    if split[i] == pair[0] and split[i + 1] == pair[1]:
                        split[i] = merge
                        del split[i + 1]
                    else:
                        i += 1

        splits = [s + [" "] for s in splits]
        splits[-1].pop()
        return [
            self.tokens.inverse.get(letter, -1) for split in splits for letter in split
        ]

    def tokens_as_strings(self, tokens: list[TokenID]) -> list[Token]:
        return [self.tokens.get(token, "UNK") for token in tokens]

    @staticmethod
    def train(lines: Iterator[str], **kwargs) -> AbstractTokenizer:  # pyright: ignore
        vocab_size = kwargs["vocab_size"]
        log.info("Training tokenizer")
        log.info("Processing data")
        # Compute word frequencies
        freqs = Counter()
        for line in tqdm(lines):
            freqs.update(get_words(line))

        # Initial vocabulary and words splits
        vocab = set([letter for word in freqs for letter in word])
        splits = {word: [c for c in word] for word in freqs}
        log.info("Done processing data, computing merges")

        # Record list of merges to build the tokenizer
        merges: Merges = {}

        start = perf_counter()
        # stats = open("stats.csv", "w")
        stats = None
        progress = tqdm()
        while len(vocab) < vocab_size:
            progress.update()
            progress.set_description(f"Vocab size: {len(vocab)}")
            pair_freqs = compute_pair_freqs(splits, freqs)
            if not pair_freqs:
                break

            most_common_pair, _ = pair_freqs.most_common(1)[0]
            merges[most_common_pair] = "".join(most_common_pair)
            log.debug(f"Merging {most_common_pair}")
            a, b = most_common_pair

            joined = "".join(most_common_pair)

            # Update splits
            for word, split in splits.items():
                if len(split) == 1:
                    continue

                i = 0
                while i < len(split) - 1:
                    if split[i] == a and split[i + 1] == b:
                        split[i : i + 2] = [joined]
                    else:
                        i += 1
                splits[word] = split

            # Could also skip this and leave unused tokens in the vocab, and instead just add `joined`
            if not APPROXIMATE_VOCAB_SIZE:
                vocab = set()
                for split in splits.values():
                    vocab.update(split)
            else:
                vocab.add(joined)

            log.debug(f"Vocabulary: {vocab}")
            if stats is not None and len(vocab) % 100 == 0:
                stats.write(f"{len(vocab)},{perf_counter() - start},{len(merges)}\n")
        if stats is not None:
            stats.close()

        vocab.add(" ")
        tokens = sorted(vocab)
        progress.close()
        log.info(f"Stopped after {progress.n} iterations, with {len(tokens)} tokens")
        return Tokenizer(merges, tokens=bidict(enumerate(tokens)))

File: src/bpe.rs
use std::io::Write;
use std::path::Path;

use itertools::Itertools;
use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet};
use serde::{Deserialize, Serialize};
use tracing::*;

use crate::{Counter, Token, TokenID, Word, UNK, UNK_STR};

type IndexMap<K, V> = indexmap::IndexMap<K, V, rustc_hash::FxBuildHasher>;
type BiMap<K, V> = bimap::BiHashMap<K, V, rustc_hash::FxBuildHasher>;

type Pair = (String, String);
type Splits<'a> = HashMap<&'a Word, Vec<Token>>;
type Merges = IndexMap<Pair, String>;

const APPROXIMATE_VOCAB_SIZE: bool = false;

fn most_frequent_pair<'a>(
    splits: &'a Splits,
    word_freqs: &'a Counter<Word>,
) -> Option<(&'a Token, &'a Token)> {
    let mut freqs = Counter::<(&String, &String)>::default();
    for (word, freq) in word_freqs.iter() {
        let split = &splits[word];
        if split.len() < 2 {
            continue;
        }
        for p in split.iter().tuple_windows::<(&String, &String)>() {
            freqs.increment(p, *freq);
        }
    }
    freqs.most_common().copied()
}

fn normalize_line(line: &str) -> String {
    line.chars()
        .filter(|c| c.is_alphanumeric() || c.is_whitespace())
        .map(|c| c.to_ascii_lowercase())
        .collect()
}

#[derive(Deserialize, Serialize)]
pub struct Tokenizer {
    merges: Merges,
    pub tokens: BiMap<TokenID, Token>,
}
impl std::fmt::Display for Tokenizer {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(
            f,
            "Tokenizer with {} tokens and {} merges",
            self.tokens.len(),
            self.merges.len()
        )
    }
}

impl crate::Tokenizer for Tokenizer {
    fn save(&self, path: impl AsRef<Path>) -> anyhow::Result<()> {
        Ok(std::fs::write(path, bincode::serialize(&self)?)?)
    }
    fn load(path: impl AsRef<Path>) -> anyhow::Result<Self> {
        Ok(bincode::deserialize(&std::fs::read(path)?)?)
    }
    fn tokens_as_string<T: IntoIterator<Item = TokenID>>(
        &self,
        // TODO: Cow
        tokens: T,
    ) -> impl Iterator<Item = String> {
        tokens.into_iter().map(|t_id| {
            self.tokens
                .get_by_left(&t_id)
                .cloned()
                .unwrap_or(UNK_STR.into())
        })
    }
    fn tokenize(&self, line: &str) -> impl Iterator<Item = TokenID> {
        // TODO: Avoid allocating
        let line = normalize_line(line);
        let mut splits: Vec<Vec<Token>> = line
            .split_whitespace()
            .map(|word| {
                word.chars()
                    .map(String::from)
                    .chain(std::iter::once(String::from(" ")))
                    .collect()
            })
            .collect();
        for (pair, merge) in &self.merges {
            for split in &mut splits {
                let mut i = 0;
                while i < split.len() - 1 {
                    if split[i] == pair.0 && split[i + 1] == pair.1 {
                        // TODO: Avoid cloning
                        split.splice(i..=i + 1, [merge.clone()]);
                    } else {
                        i += 1;
                    }
                }
            }
        }
        splits
            .into_iter()
            .flatten()
            .map(|token| self.tokens.get_by_right(&token).copied().unwrap_or(UNK))
    }
    fn train(lines: impl IntoIterator<Item = String>, vocab_size: usize) -> Self {
        let start = std::time::Instant::now();
        info!("Training tokenizer");
        info!("Processing data");
        let mut word_freqs = Counter::<Word>::default();
        for line in lines {
            for word in normalize_line(&line).split_whitespace() {
                // TODO: without allocating?
                word_freqs.insert(word.into());
            }
        }
        let mut vocab =
            HashSet::<Token>::with_capacity_and_hasher(vocab_size, rustc_hash::FxBuildHasher);
        vocab.extend(word_freqs.keys().flat_map(|w| w.chars()).map(Token::from));
        let mut splits: Splits = word_freqs
            .keys()
            .map(|word| (word, word.chars().map(Token::from).collect()))
            .collect();
        info!("Done processing data, computing merges");

        let mut merges = Merges::with_capacity_and_hasher(vocab_size, rustc_hash::FxBuildHasher);
        let progress = indicatif::ProgressBar::no_length();
        progress.set_style(
            indicatif::ProgressStyle::with_template("[{elapsed_precise}] {msg}   {per_sec}")
                .unwrap(),
        );
        let mut stats: Option<std::io::BufWriter<std::fs::File>> = None;
        // let mut stats = std::fs::File::create("stats.csv")
        //     .inspect_err(|e| warn!("Failed to create stats file: {}", e))
        //     .ok()
        //     .map(std::io::BufWriter::new);
        while vocab.len() < vocab_size {
            progress.set_message(format!("Vocab size: {}", vocab.len()));
            progress.inc(1);
            let Some(pair) =
                most_frequent_pair(&splits, &word_freqs).map(|(a, b)| (a.clone(), b.clone()))
            else {
                break;
            };
            let joined: String = pair.0.clone() + &pair.1;
            debug!("Merging {:?}", pair);
            // Update splits
            for split in splits.values_mut() {
                if split.len() == 1 {
                    continue;
                }
                let mut i = 0;
                while i < split.len() - 1 {
                    if pair.0 == split[i] && pair.1 == split[i + 1] {
                        split.splice(i..=i + 1, [joined.clone()]);
                    } else {
                        i += 1;
                    }
                }
            }
            // Update vocab and merges
            if APPROXIMATE_VOCAB_SIZE {
                vocab.insert(joined.clone());
            }
            merges.insert(pair, joined);
            if !APPROXIMATE_VOCAB_SIZE {
                vocab = splits.values().flatten().map(Token::from).collect();
            }
            if vocab.len() % 100 == 0 {
                if let Some(stats) = &mut stats {
                    stats
                        .write_all(
                            format!(
                                "{},{},{}\n",
                                vocab.len(),
                                start.elapsed().as_secs_f32(),
                                merges.len()
                            )
                            .as_bytes(),
                        )
                        .unwrap();
                }
            }
        }
        progress.finish();

        vocab.insert(" ".into());
        let tokens: BiMap<_, _> = vocab.into_iter().sorted().enumerate().collect();
        info!(
            "Stopped after {} iterations, with {} tokens",
            progress.position(),
            tokens.len()
        );
        info!(elapsed = ?start.elapsed(), "Done training tokenizer");
        Self { merges, tokens }
    }
}
